### ISSUE TITLE
[FEATURE] Registering an overpass-api knitr engine

### ISSUE THREAD

**Issue by idshklein** [ropensci/osmdata#231]
The use of overpass-api is a daily routine for many R developers. I usually write overpass queries using https://overpass-turbo.eu/. However, it occurred to that it is possible to write overpass-api query straight from Rmarkdown files. 
Following [these](https://bookdown.org/yihui/rmarkdown-cookbook/custom-engine.html) instructions, I was able to retrieve the output of an overpass-api query, yet I find it difficult to pass the query and retrieve a mapview object of the result. 
Reprexing an rmarkdown flow is somewhat difficult, and due to the fact this is a POC, I'll show my results in pngs. 

This worked:
![image](https://user-images.githubusercontent.com/12258810/103179720-6f547680-4897-11eb-8407-ebcc1e0faedd.png)

This didn't: 
![image](https://user-images.githubusercontent.com/12258810/103179749-cd815980-4897-11eb-8e88-5083a546f724.png)

I heavily rely on the parsing functions from the osmdata to try and create the map output, so I thought it would be best if this was a part of the osmdata package. 
This is still in very premature development, as you can see, the query can't have comments or new lines, otherwise it does not get parsed. However I though this might interest the developers of osmdata, due to this feature added value and comfortably. 

Many more options can be added to this feature, such as output type, overpass_url, color parsing of the query code such as in the png below(taken from https://overpass-turbo.eu/), and many more. 
![image](https://user-images.githubusercontent.com/12258810/103179849-b858fa80-4898-11eb-95a0-e11e1191033b.png)

**Comment by mpadge** [comment]
Thanks @idshklein, this does indeed look like an interesting extension for `osmdata`. I nevertheless remain somewhat unsure exactly what you are attempting to do. I can see the advantage of having an engine to render symtax-highlighted queries, but can't really see the advantage of using a custom engine to insert a `mapview` object in comparison with direct usage of `mapshot`? Could you perhaps explain a little better both your motivation here, and how the intended results would differ from direct `mapshot` usage? Or even better, the more general advantages of using a custom `knitr` engine for the `overpass-api`? Oh, and could you also please ensure that you paste at least some direct code in the issue, so i can easily copy-paste it to recreate what you're trying to do?

Beyond that, there are obviously a number of other issues which your comments raised, as you state. From a package developer side, note that an aim of the `osmdata` package is to enable users as much as possible to avoid having to formulate their own `overpass` queries. In your case, you are using the `area` filter, which obviously necessitates a custom query in current form. It would nevertheless be very straightforward to implement an `opq_area` function so you could just call
``` r
opq_area(id = 6195356, type = "relation") %>%
  add_osm_feature(key = "amenity", value = "fuel") %>%
  osmdata_sf() # or whatever
```
Feel free to open a separate issue on that, or even better submit a PR. Looking forward to seeing where this issue goes ... thanks for the input

**Comment by idshklein** [comment]
I may have been a bit unclear, sorry for that. 
The main idea is to create a similar engine to [](https://overpass-turbo.eu/), In order to ease query exploration. I opened this issue in osmdata because the feature I am suggesting highly depends on the methods that already exist in the library. 
Inserting a mapview allows for the exploration of the results, just like in [](https://overpass-turbo.eu/).
The general advantage of using a knitr custom engine in order to be able to embed the results in rmarkdown, is allowing the sharing of results in a way that can have context, instead of just sending a link to a query. I do not know other ways than knitr to do it in rmarkdown, and I'll be grateful if you can point out to other options if you know them. 
I'm sorry for the screenshots, I should have send code. it is attached at the end of this comment. It's just that it's hard sending repexes in an rmarkdown environment, wasn't sure what I was supposed to send. 
The opq_area function is a good Idea, and I'll try to find time writing an issue about it before I submit a PR, with sample code. 

```{r setup, echo=TRUE}
knitr::opts_chunk$set(echo = TRUE)
library(osmdata)
library(mapview)
```
This worked (expected an xml output and got it):
```{r}
knitr::knit_engines$set("overpass-api" = function(options) {
  code <- paste(options$code, collapse = "")
  res <- osmdata:::overpass_query(txt)
  if(options$eval) {
    output <- xml2::read_xml(res)
  }
  else{
    output <- code
  }
  knitr::engine_output(options, code, output)
})
```
```{overpass-api echo=TRUE}
[out:json][timeout:25];(node["amenity"="fuel"](area:3601246791);way["amenity"="fuel"](area:3601246791);relation["amenity"="fuel"](area:3601246791););out body;>;out skel qt;
```

This didn't (expected a mapview and didn't get it):
```{r}
knitr::knit_engines$set("overpass-api" = function(options) {
  code <- paste(options$code, collapse = "")
  res <- osmdata:::overpass_query(txt)
  xml2::write_xml(xml2::read_xml(res), tmp <- tempfile(fileext = "xml"))
  sf_res <- osmdata_sf(doc = tmp)
  sf_res_1 <-  purrr::map(sf_res[stringr::str_starts(names(sf_res),"osm")],function(x){
    if(is.null(x) || nrow(x)== 0){
      return(NULL)
    }else {
      return(x)
    }
  })
  map <- mapview(sf_res_1$osm_points) + mapview(sf_res_1$osm_lines) + mapview(sf_res_1$osm_polygons) + mapview(sf_res_1$osm_multilines) + mapview(sf_res_1$osm_multipolygons)
  if(options$eval) {
    output <- map
  }
  else{
    output <- xml2::read_xml(tmp)
  }
  knitr::engine_output(options, code, output)
})
```
```{overpass-api echo=TRUE}
[out:json][timeout:25];(node["amenity"="fuel"](area:3601246791);way["amenity"="fuel"](area:3601246791);relation["amenity"="fuel"](area:3601246791););out body;>;out skel qt;
```

**Comment by idshklein** [comment]
I found some sort of solution, however it is somewhat ugly, and should be further developed. 
the result is here:
https://idshklein.github.io/overpass-api-rmarkdown/intial_attempt.html
the code is here:
https://github.com/idshklein/overpass-api-rmarkdown

**Comment by mpadge** [comment]
Thanks @idshklein, now i see from your examples what you're getting at. That is indeed really impressive! Feel free to re-open this issue if you're interesting in integrating some of those ideas within this package. My suggestion for a start would be to translate what you've done into a vignette which demonstrates how to do it. Once we had the vignette, a second step could then be to translate that into a `knitr::engine_set` call within the package itself. As said above, please also feel free to open a separate issue on a direct `opq_area()` function so that your examples can avoid calling overpass directly. I'd be happy to do that if you'd prefer not to - just let me know. Either way, that function should be done first before addressing any of the `knitr` stuff. Thanks again for raising these ideas to begin with, and for following up on them!

---

An aside: I'm based in Muenster, Germany, which just so happens to be a sister city of Rishon Lezion. Nice coincidence!
